### ISSUE TITLE
Change behaviors of rescue_from to not depend on mount order

### ISSUE THREAD

**Issue by wowinter13** [ruby-grape/grape#1975]
Hi, found an interesting case when `rescue_from :all` should override outer `rescue_from :error_class`
```ruby
class PublicApi < Grape::API
  rescue_from ActiveRecord::RecordNotFound do
    binding.pry
    error_response(status: 403)
  end
  mount AuthController
end
class AuthController < Grape::API
  rescue_from :all do 
    error_response(status: 403)
  end

  post :request do
    User.find_by!(phone: phone)
  end
end
```
```bash
15:  rescue_from ActiveRecord::RecordNotFound do
15:     binding.pry
16:     error_response(status: 403)
17:  end
```
For some reason, it seems to not working correctly. I suppose `AuthController rescue_from :all` to be called first.
But if we mount AuthController above `rescue_from ActiveRecord::RecordNotFound` it would work correctly.
So, does it look like a bug or additional info to documentation?

**Comment by dblock** [comment]
Can you turn it into a spec? Looks incorrect to me.

**Comment by wowinter13** [comment]
@dblock Should I create a PR with spec? Or gist in the discussion would be enough?

**Comment by dblock** [comment]
> @dblock Should I create a PR with a spec? Or gist in the discussion would be enough?

Yes. And don't hesitate to try a fix!

**Comment by jcagarcia** [comment]
Hey guys! 👋 

I've been examining this issue and I'm uncertain if it qualifies as a bug. As far as I can tell, Grape seems to be functioning as intended.

In your `PublicApi < Grape::API` class, you've specified a `rescue_from` for the specific class `ActiveRecord::RecordNotFound`. Following this, you're mounting the `AuthController < Grape::API` class, which has a declared `rescue_from :all`.

What's happening here is that upon mounting, you essentially "merge" the `AuthController` into the `PublicApi` context, inheriting all the defined `rescue_from` methods. Nevertheless, your primary class remains `PublicApi`.

Consequently, when an error occurs, the first error handlers considered are those within the class. If none apply, it then resorts to the `:all` handlers. 

https://github.com/ruby-grape/grape/blob/3901bf4ddc85b83bf6df9c26b6850f2d6117a663/lib/grape/middleware/error.rb#L43-L46

In your scenario, the initial entry point is `rescue_from ActiveRecord::RecordNotFound` because even though you're invoking an endpoint defined in the `AuthController` class, the context remains within the `PublicApi` class. Hence, the initially considered handler is `rescue_from ActiveRecord::RecordNotFound`.

By altering the order of mounting before the `rescue_from ActiveRecord::RecordNotFound` definition, the rescue handler isn't in place when those endpoints are mounted. Consequently, as @wowinter13  correctly said, the `:all` rescue is directly invoked. This is why the sequence impacts the handler chosen.

With that being said, I still maintain the view that this behavior is not a bug; rather, it aligns with the expected functionality.

What do you think? Make sense?

**Comment by dblock** [comment]
Thanks for the deep dive! If you compare this behavior with `version`, then it looks different: a `version` declaration in `AuthController` would take precedence over the declaration in `PublicApi`, won't it? In that case logically at least we'd want locality to be consistent? WDYT?

**Comment by dblock** [comment]
I think you're completely correct on all the points above. 

So, is the problem in this issue correctly formulated or do we need to edit it? What's the tl;dr of the ask? What do we intend to change?

Either way, before changing anything, we have to start by ensuring that we have specs covering all current behavior and documentation that expresses it correctly. Appreciate any PRs!

**Comment by jcagarcia** [comment]
Hey @dblock , 

I think that the issue should be edited from the original title "An incorrect behavior of :rescue_from" to something like "Different behaviors of rescue_from method when changing mount order"

Let me work on this :)

**Comment by jcagarcia** [comment]
Hi again @dblock ,

I was working on documenting the current behavior and to add some tests for these scenarios and I found that there is already an explanation about the inherited `rescue_from` handlers and the `mount` order.

https://github.com/ruby-grape/grape/blob/3901bf4ddc85b83bf6df9c26b6850f2d6117a663/README.md?plain=1#L411-L422

In fact, this was already managed in #646 some years ago.

So my question now is... do we want to change how the inheritance works when using `mount`?

**Comment by dblock** [comment]
I renamed this to "Change behaviors of rescue_from to not depend on mount order", lmk if this can be improved.

> So my question now is... do we want to change how the inheritance works when using `mount`?

Well, no. Personally the order is predictable and logical (once you understand it). But the OP, @wowinter13 was visibly confused, so it tells me that I could be wrong. I guess I'd be looking for some strong opinions from various people and code contributions that change the behavior! Happy to be the arbiter.

**Comment by jcagarcia** [comment]
Hey @dblock ,

After taking a deep look to this, I think I have a possible solution that seems to be workings.

Take a look to the explanation of the PR https://github.com/ruby-grape/grape/pull/2384

**Comment by jcagarcia** [comment]
This has been solved at #2384 :)

**Comment by dblock** [comment]
Maybe @wowinter13 wants to try it out from HEAD?
